src/rust/engine/src/rule_graph.rs

```diff
@@ -794,6 +849,35 @@ impl RuleEdges {
     }
     RuleEdges { dependencies: new_deps, selector_to_dependencies: new_selector_deps }
   }
+
+  pub fn entries_for(&self, selectors: &Vec<Selector>) -> Entries {
+    let r = self.selector_to_dependencies.get(selectors).map(|x|x.clone()).unwrap_or_else(|| Vec::new());
+
+    if false {
+      println!("genning for {:?} found {}. total avail {}", selectors.iter().map(|x| selector_str(x)).collect::<Vec<_>>(), r.len(), self.dependencies.len());
+      if selectors.len() > 1 {
+        println!("{}",
+                 self.selector_to_dependencies
+                   .keys()
+                   .map(|ss|
+                     ss.iter()
+                       .map(|s| selector_str(s))
+                       .collect::<Vec<_>>()
+                       .join(", "))
+                   .collect::<Vec<_>>()
+                   .join("\n   __ ")
+        );
+
+        let w_first = self.selector_to_dependencies.iter()
+          .filter(|&(k, _)| k.len() >= 1 && k[0] == selectors[0])
```

Comment:
I generally prefer `!coll.is_empty()` to `coll.len() >= 1` (len is probably `O(1)` here, but it still sends a cleaner signal).